analyzers/src/SonarAnalyzer.CSharp/Rules/UseConstantLoggingTemplate.cs

```diff
@@ -0,0 +1,130 @@
+        new LogMethod(KnownType.Serilog_Log, "Fatal", "messageTemplate"),
+        new LogMethod(KnownType.Serilog_Log, "Information", "messageTemplate"),
+        new LogMethod(KnownType.Serilog_Log, "Verbose", "messageTemplate"),
+        new LogMethod(KnownType.Serilog_Log, "Warning", "messageTemplate"),
+        new LogMethod(KnownType.log4net_ILog, "Debug", "message"),
+        new LogMethod(KnownType.log4net_ILog, "DebugFormat", "format"),
+        new LogMethod(KnownType.log4net_ILog, "Error", "message"),
+        new LogMethod(KnownType.log4net_ILog, "ErrorFormat", "format"),
+        new LogMethod(KnownType.log4net_ILog, "Fatal", "message"),
+        new LogMethod(KnownType.log4net_ILog, "FatalFormat", "format"),
+        new LogMethod(KnownType.log4net_ILog, "Info", "message"),
+        new LogMethod(KnownType.log4net_ILog, "InfoFormat", "format"),
+        new LogMethod(KnownType.log4net_ILog, "Trace", "message"),
+        new LogMethod(KnownType.log4net_ILog, "TraceFormat", "format"),
+        new LogMethod(KnownType.log4net_ILog, "Warn", "message"),
+        new LogMethod(KnownType.log4net_ILog, "WarnFormat", "format"),
+        new LogMethod(KnownType.Castle_Core_Logging_Ilogger, "Debug", "message"),
+        new LogMethod(KnownType.Castle_Core_Logging_Ilogger, "DebugFormat", "format"),
+        new LogMethod(KnownType.Castle_Core_Logging_Ilogger, "Error", "message"),
+        new LogMethod(KnownType.Castle_Core_Logging_Ilogger, "ErrorFormat", "format"),
+        new LogMethod(KnownType.Castle_Core_Logging_Ilogger, "Fatal", "message"),
+        new LogMethod(KnownType.Castle_Core_Logging_Ilogger, "FatalFormat", "format"),
+        new LogMethod(KnownType.Castle_Core_Logging_Ilogger, "Info", "message"),
+        new LogMethod(KnownType.Castle_Core_Logging_Ilogger, "InfoFormat", "format"),
+        new LogMethod(KnownType.Castle_Core_Logging_Ilogger, "Trace", "message"),
+        new LogMethod(KnownType.Castle_Core_Logging_Ilogger, "TraceFormat", "format"),
+        new LogMethod(KnownType.Castle_Core_Logging_Ilogger, "Warn", "message"),
+        new LogMethod(KnownType.Castle_Core_Logging_Ilogger, "WarnFormat", "format"));
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+        {
+            var invocation = (InvocationExpressionSyntax)c.Node;
+            var candidates = LogMethods.Where(x => x.MethodName == invocation.GetName()).ToArray();
+            if (candidates.Any()
+                && c.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol method
+                && Array.Exists(candidates, x =>  x.ContainingType.Matches(method.ContainingType))
+                && method.Parameters.FirstOrDefault(x => Array.Exists(candidates, c => c.MessageParameterName == x.Name)) is { } messageParameter
```

Comment:
`c` hides the `c` variable from the main action and may create confusion. I think another name should be used.